### ISSUE TITLE
xreflabel for bibliography anchor is ignored

### ISSUE THREAD

**Issue by getreu** [asciidoctor/asciidoctor#1562]
## Please consider

![test adoc](https://cloud.githubusercontent.com/assets/579082/10881971/c6a8ad4c-816f-11e5-9181-2e822569c946.png)
## This is the expected rendition
- ![test xml](https://cloud.githubusercontent.com/assets/579082/10881981/d5ce52d6-816f-11e5-9c09-d9e926cb8c9d.png)
- please note the little blue `3` 
  ![asciidoctor-fopub](https://cloud.githubusercontent.com/assets/579082/10881984/db40cec4-816f-11e5-9f5c-e534230e6d09.png)
## asciidoctor does not honor `xreflabel`

Note that [goethe] should should be substituted by `xreflable` (here 3) but it isn't! 
![asciidoctor2html](https://cloud.githubusercontent.com/assets/579082/10882015/15fb0052-8170-11e5-99bf-5805fef3a7c8.png)
## neither asciidoctor-pdf does

![asciidoctor-pdf](https://cloud.githubusercontent.com/assets/579082/10882032/387bdeb2-8170-11e5-922f-11314605f186.png)

**Comment by mojavelinux** [comment]
Here's the AsciiDoc in plain text for reference:

```
<<goethe>>

[bibliography]
== References

[[goethe,3]][3]:: Die neuen Leiden des Jungen Werther.
```

**Comment by mojavelinux** [comment]
Technically, you are not writing this correctly. You should be writing:

```
<<goethe>>

[bibliography]
== References

[bibliography]
- [[goethe,[3]{zwsp}]][3] Die neuen Leiden des Jungen Werther.
```

The `{zwsp}` is necessary to prevent the end square bracket from getting confused from the square brackets of the enclosure.

I'll admit, though, that the bibliography is, in general, very inflexible. It has a very opinionated way of defining entries and I think we need to extend that.

The main problem here is that the bibliography anchor (triple square brackets) doesn't support an xreflabel, so we're technically using the normal example to get this example to work. So this is just a workaround.

Here's what needs to be changed:

1. The bibliography anchor needs to support the xreflabel. It should also accommodate a closing square bracket in that label as part of the content.
2. A reference to a bibliography anchor needs to honor the xreflabel value.

**Comment by mojavelinux** [comment]
It would be helpful if you provided an example of how you want it to look. Feel free to edit the HTML so that you can take a screenshot to show how you want it to appear.

Here's the syntax that should work (once the fix is made):

```
<<goethe>>

[bibliography]
== References

[bibliography]
- [[[goethe,[3]]]] Die neuen Leiden des Jungen Werther.
```

The second `[3]` shouldn't be necessary since the xreflabels should be used at the anchor point as well.

**Comment by mojavelinux** [comment]
I've updated the title to describe the problem more accurately. We need to deal with the behavior in Asciidoctor PDF separately. We can follow up with an issue there once the problem in core is resolved.

**Comment by mojavelinux** [comment]
Btw, that's for reporting this!

**Comment by mojavelinux** [comment]
An alternative to the `{zwsp}` hack for the closing bracket is to use a numeric character reference:

```
[[goethe,&#91;3&#93;]][3]
```

**Comment by getreu** [comment]
Thank you for the workaround! `[[[goethe,[3]]]]` looks fine to me.

Many bibliography styles uses continuous numbering, e. g. IEEE

```
According to [<<Anderson>>, p. 132] ....

[bibliography]
== References

{counter2:bib:0}
[horizontal]

[[Anderson,{counter:bib}]][{bib}]:: B. Anderson, L. Bergstrom, D. Herman, J.
Matthews, K. McAllister, M.  Goregaokar, J. Moffitt, and S. Sapin, ‘Experience
Report: Developing the Servo Web Browser Engine using Rust’, _arXiv preprint
arXiv:1505.07383_, 2015.
```
Will the first line then look like this?
```
[[[Anderson,{counter:bib}]]]:: B. Anderson, L. Bergstrom, D. Herman, J. ...
```

**Comment by mojavelinux** [comment]
> Will the first line then look like this?

In this case, it would be best to have an auto-counter.

> Many bibliography styles uses continuous numbering, e. g. IEEE

The main problem with the current bibliography is that it isn't really a bibliography and not very well thought out. The goal here is to design it the way it should be so that it does something more sensible by default.

We'll also need to consider where [asciidoctor-bibtex](https://rubygems.org/gems/asciidoctor-bibtex) fits in. My feeling is that Asciidoctor should at least produce something sensible by default and allow the use of asciidoctor-bibtex for more advanced usage and automatic lookups.

**Comment by getreu** [comment]
The suggested {zwsp} hack does not help in my case, because [] should not be part of xreflabel anyway: 

    According to [<<goethe>>, p. 32] 

The [] brackets often include additional information like page numbers as showed above.

Quoting you: 

>    The bibliography anchor needs to support the xreflabel. ~~It should also accommodate a closing square bracket in that label as part of the content.~~
>    A reference to a bibliography anchor needs to honor the xreflabel value.

**Comment by mojavelinux** [comment]
Thanks for the clarification.

**Comment by getreu** [comment]
Consider: 

```
[[Avižienis04,{counter:bib}]] A. Avižienis, J.-C. Laprie, B. Randell, and
C. Landwehr, ‘Basic concepts and taxonomy of dependable and secure computing’,
_Dependable and Secure Computing, IEEE Transactions on,_ vol. 1, no. 1, pp.
11–33, 2004.
```

Tags like `Avižienis04` should allow Unicode if possible. This is especially important for Greek, Arab and Cyrillic names.

**Comment by mojavelinux** [comment]
Of course, Asciidoctor supports character ranges across the entire unicode range. In this case, the name must consist of word characters (in addition to hyphens, periods and colons). If a glyph is categorized as a word character by Unicode, it will be accepted.

You don't need to use the `counter2`. Instead, you can (and should) just use an attribute entry for `bibtex`, which has the benefit that it won't take up space in the output.

Here's my recommendation:

```
= Document Title

According to <<Avižienis,[Avižienis, p. 32]>>...

[bibliography]
== Bibliography

:bibidx: 0
[bibliography]
* [[Avižienis,{counter:bibidx}]][{bibidx}] A. Avižienis, J.-C. Laprie, B. Randell, and C. Landwehr,
‘Basic concepts and taxonomy of dependable and secure computing’,
_Dependable and Secure Computing, IEEE Transactions on,_ vol. 1, no. 1, pp. 11–33, 2004.
```

Unfortunately, the number (bibidx) isn't going to get inserted at the source location because the parsing just starts to break down at this point (for various reasons). We have to move to a more advanced system to get that level of control. We're working with what we've got here.

One thing I noticed is that inline anchors in the first line of a list element are not being cataloged. You can force it to be cataloged using a fake line:

```
[bibliography]
* {blank}
[[Avižienis,{counter:bibidx}]][{bibidx}] A. Avižienis, J.-C. Laprie, B. Randell, and C. Landwehr,...
```

However, by doing that we see substitutions are not applied to the xreflabel at the source location. For instance, if you use `<<Avižienis>>` at the source location, it ends up looking like:

```
{counter:bib}{zwsp}
```

It's abundantly clear to me at this point that asciidoctor-bibtex (or something like it) is far more appropriate for handling this information than the low-level, built-in constructs in AsciiDoc. We shouldn't be typing all this erroneous markup in the AsciiDoc document when it can be worked out automatically by using an explicit notation system.

**Comment by getreu** [comment]
I checked out your document example. As long as there is a  `ž` in the anchor's name I can not reference it . Once changed to `z` it works, even without `* {space}`.

**Comment by mojavelinux** [comment]
I don't understand what you mean by "reference it". I'll need a simple document that shows the whole example that you are observing does not work. Also, please let me know what Asciidoctor processor you are using.

```
asciidoctor -v
```

Thanks!

**Comment by mojavelinux** [comment]
Are you observing this behavior in an IDE, like AsciidocFX?

**Comment by getreu** [comment]
First without accent over `z`.

```
$ asciidoctor -v
Asciidoctor 1.5.4.dev [http://asciidoctor.org]
Runtime Environment (ruby 2.1.5p273 (2014-11-13) [x86_64-linux-gnu]) (lc:UTF-8 fs:UTF-8 in:- ex:UTF-8)
```
All tools are executed from command line. 


[20160104-Document Title.adoc.txt](https://github.com/asciidoctor/asciidoctor/files/77123/20160104-Document.Title.adoc.txt)

![20160104-document title html](https://cloud.githubusercontent.com/assets/579082/12088980/a1a0b6ca-b2e8-11e5-921c-22660c8575a4.png) - not ok


[20160104-Document Title  asciidoctor-fopub.pdf](https://github.com/asciidoctor/asciidoctor/files/77119/20160104-Document.Title.asciidoctor-fopub.pdf) - OK! 

[20160104-Document Title  asciidoctor-pdf.pdf](https://github.com/asciidoctor/asciidoctor/files/77120/20160104-Document.Title.asciidoctor-pdf.pdf) - not ok

**Comment by getreu** [comment]
Finally Avižienis and Avizienis compiles the same. It just messed up my syntax highlighter.
Sorry, my mistake!

What annoys me at the moment: I have to finish an article and the editor expects me to send it as .docx. 
The best way I found so far is to:
1. first produce an .html, 
2. then import it into LibreOffice and then 
3. export as .docx.

Unfortunately 1. the .html prints [[Avizenis], p. 2] instead of [1, p. 2]. The .pdf output with asciidoctor-fopub would be ok, but it does not meed my editors requirements...

**Comment by mojavelinux** [comment]
Thanks for digging into it. I'm glad to hear it was just the syntax highlighting. While the syntax highlighting is helpful, it often gets confused in my experience...and as this example shows.

If you have the time, you might want to look into the docx converter that was created by Greg Turnquist. See https://github.com/gregturn/asciidoctor-packt. The styling is specific for packt, but it's a good starting point for creating a custom docx file.

Another option would be converting to DocBook, then converting to docx using pandoc.

**Comment by getreu** [comment]
Thank you for the hint! I will definitely check this out for the next article. I am late for this one...

For completeness: it seems it is the chrome-plugin that has this problem with accents in labels:

.no  ž in label 
![no  ž in label](https://cloud.githubusercontent.com/assets/579082/12089611/79308ede-b2ef-11e5-9946-137e3d372c76.png)

.with ž in label
![with ž in label](https://cloud.githubusercontent.com/assets/579082/12089624/a6e05cc4-b2ef-11e5-803f-adcdd2fc6f95.png)

**Comment by mojavelinux** [comment]
Just as I suspected, you're seeing different results in different environments. While Asciidoctor core (Ruby) and AsciidoctorJ have complete support for Unicode, Asciidoctor.js does not. That's because the JavaScript regular expression is primitive and doesn't provide proper Unicode support. (even in JavaScript 6).

Fortunately, there's a workaround that we can implement for Asciidoctor.js, but we have not yet done it. See https://github.com/asciidoctor/asciidoctor.js/issues/43. Until then, any part of the syntax that relies on matching a word or letter character specifically is not going to work outside of basic Latin.

For a great article on JavaScript's Unicode problem, see https://mathiasbynens.be/notes/javascript-unicode

**Comment by mojavelinux** [comment]
To clarify, the following returns false:

```
/^[A-Za-z]+$/.test('Avižienis')
```

while the following returns true:

```
XRegExp('^\\pL+$').test('Avižienis')
```

**Comment by getreu** [comment]
## Workaround workflow

Here my workaround to get the same output for both workflows:  

### asciidoctor(adoc) = html; post_process_html(html) = html
   
  `asciidoctor -r asciidoctor-diagram -b xhtml5 myfile.adoc`
 `./post_process_html myfile.html"`


### asciidoctor(adoc) = xml, asciidoctor-fopub(xml) =pdf  

   `asciidoctor -b docbook  -a data-uri! "myfile.adoc"`
   `asciidoctor-fopub/fopub "myfile.xml"`


## Workaround script

File: `/post_process_html`

```
#!/bin/sh

Input="$1"


sed -i "$Input" \
	-e 's|\[Weinstein\]|1|g' \
	-e 's|\[Schneier\]|2|g' \
	-e 's|\[Eikenberg\]|3|g' \
	-e 's|\[Bongard\]|4|g' \
	-e 's|\[Greenberg\]|5|g' \
	-e 's|\[Kopetz\]|6|g' \
	-e 's|\[Avizienis\]|7|g' \
	-e 's|\[VDC\]|8|g' \
	-e 's|\[Szekeres\]|9|g' \
	-e 's|\[Durumeric\]|10|g' \
	-e 's|\[Mutton\]|11|g' \
	-e 's|\[Codenomicon\]|12|g' \
	-e 's|\[Riley\]|13|g' \
	-e 's|\[Riley2\]|14|g' \
	-e 's|\[Bocek\]|15|g' \
	-e 's|\[Venafi\]|16|g' \
	-e 's|\[Tuexen\]|17|g' \
	-e 's|\[OpenSsl\]|18|g' \
	-e 's|\[Hern\]|19|g' \
	-e 's|\[CWE\]|20|g' \
	-e 's|\[Timson\]|21|g' \
	-e 's|\[NL\]|22|g' \
	-e 's|\[Cimpanu\]|23|g' \
	-e 's|\[Rust\]|24|g' \
	-e 's|\[Rust-Benchmark\]|25|g' \
	-e 's|>Dependability and security attributes</a|>Figure 1</a|g' \
	-e 's|>TLS Heartbeat protocol</a|>Figure 2</a|g' \
	-e 's|>TLS Heartbeat with spoofed payload-length-field</a|>Figure 3</a|g' \
	-e 's|>Heartbleed patch</a|>Figure 4</a|g' \
	-e 's|>Development Concepts</a|>Figure 5</a|g' \
	-e 's|>Memory layout of a Rust vector</a|>Figure 6</a|g' \
	-e 's|>Memory layout of a Java vector</a|>Figure 7</a|g' \
	-e 's|>The Heartbleed Vulnerability</a|>the section called “The Heartbleed Vulnerability”</a|g' \
	-e 's|>Could Heartbleed Have Happened With Rust?</a|>the section called “Could Heartbleed Have Happened With Rust?”</a|g' \
	-e 's|>Rust on Embedded Systems</a|>the section called “Rust on Embedded Systems”</a|g' \
	-e 's|>Common weaknesses in C/C++ that affect memory</a|>Table 1, “Common weaknesses in C/C++ that affect memory”</a|g' \
	-e 's|>Vulnerable Heartbeat code in C</a|>Table 2, “Vulnerable Heartbeat code in C”</a|g' \
	-e 's|>Patched Heartbeat code in C</a|>Table 3, “Patched Heartbeat code in C”</a|g' \
	-e 's|>Vulnerable Heartbeat code in Rust</a|>Table 4, “Vulnerable Heartbeat code in Rust”</a|g' \
	-e 's|>Simulation of Heartbleed exploit package</a|>Table 5, “Simulation of Heartbleed exploit package”</a|g' \
	-e 's|>System response after Heartbleed attack</a|>Table 6, “System response after Heartbleed attack”</a|g' \
	-e 's|>Patched Heartbeat code in Rust</a|>Table 7, “Patched Heartbeat code in Rust”</a|g' \
	-e 's|>Ressource sharing in Rust</a|>Table 8, “Ressource sharing in Rust”</a|g' \
	-e 's|>Common weaknesses in C/C++ affecting memory avoidable with iterators</a|>Table 9, “Common weaknesses in C/C++ affecting memory avoidable with iterators”</a|g' \
	-e 's|>Vigenère cipher in Rust</a|>Table 10, “Vigenère cipher in Rust”</a|g' \
	-e 's|<h2\(.*\)/h2>|<h1\1/h1>|g' \
	-e 's|<h3\(.*\)/h3>|<h2\1/h2>|g' \

#	-e 's|<div id="preamble"\(.*\)>|<div class="Abstract"\1>|g' \
```

**Comment by getreu** [comment]
Summery of this thread:
Here's what needs to be changed : (thanks to mojavelinux)

1.    The bibliography anchor needs to support the xreflabel. (asciidoctor)
2.    A reference to a bibliography anchor needs to honor the xreflabel value. (asciidoctor + asciidoctor-pdf)
3.    Asciidoctor.js'  Unicode problem
To clarify, the following returns false:

         /^[A-Za-z]+$/.test('Avižienis')
while the following returns true:

        XRegExp('^\\pL+$').test('Avižienis')

**Comment by mojavelinux** [comment]
> Asciidoctor.js' Unicode problem

This is finally underway. See https://github.com/asciidoctor/asciidoctor.js/pull/284. No more hacks will be needed.

**Comment by mojavelinux** [comment]
I'm going to resolve this in too phases. First, I'm going to bring the behavior of bibliography anchors up to par with the normal inline anchors. The solution to #858 will bring greater flexibility that will allow you to have much more control over how the anchors are styled.

In this first phase, I will allow reftext to be specified. However, it will be enclosed in square brackets when displayed just like the ID is now. This essentially allows you to use a different label internally than what is shown.

The ability to do things like attribute references will be deferred until #858 is solved since it requires the mechanics that are being introduced in that fix. So this is an interim step.

**Comment by mojavelinux** [comment]
Here's how you'll be able to assign a numeric label to a bibliography entry:

```
[bibliography]
* [[[fowler,1]]] Martin Fowler. _Refactoring. Improving the Design of Existing Code_. 2002. p.212.
```

You can then reference it as:

```
Please read <<fowler>>.
```

That will display as:

```
Please read [1].
```

**Comment by mojavelinux** [comment]
A PR for the proposed initial solution has been sent. The next step will be to store a reference to the bibliography anchor node in the refs table of the document catalog. That will give you the ability to tailor the text used to refer to a bibliography anchor in a custom converter.

More important, it will allow extensions (such as asciidoctor bibtex) to access the collection of bibliography anchors in the document catalog so that they don't have to be scanned again. In fact, this use case helps me understand what to focus on. The core processor should give information that extensions cannot access without parsing the document again.

**Comment by mojavelinux** [comment]
The proposed solution has been merged as 7cafe291a55f7754bd25f62dae6d00ebb3444a7d.

I will continue to reference this issue as I work on #858 (& co), so this is not the end of the story. But I'm considering the original request as satisfied.
